File: Difficulty/Medium/[138] Copy List with Random Pointer/solution.cpp
/*
 * @lc app=leetcode id=138 lang=cpp
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/
class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, int> m;
        vector<Node*> record;
        int i = 0;
        Node *new_head = new Node(0);
        Node *tmp = head, *new_tmp = new_head;
        while (tmp) {
            new_tmp->next = new Node(tmp->val);
            m[tmp] = i;
            new_tmp = new_tmp->next;
            tmp = tmp->next;
            record.push_back(new_tmp);
            i++;
        }
        tmp = head;
        new_tmp = new_head;
        while (tmp) {
            if (tmp->random)
                new_tmp->next->random = record[m[tmp->random]];
            else
                new_tmp->next->random = nullptr;
            tmp = tmp->next;
            new_tmp = new_tmp->next;
        }
        return new_head->next;
    }
};
// @lc code=end

File: Difficulty/Medium/[138] Copy List with Random Pointer/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        m, record, i = dict(), [], 0
        new_head = Node(0)
        tmp, new_tmp = head, new_head
        while tmp:
            new_tmp.next = Node(tmp.val)
            m[tmp] = i
            new_tmp, tmp = new_tmp.next, tmp.next
            record.append(new_tmp)
            i += 1
        tmp = head
        new_tmp = new_head
        while tmp:
            if tmp.random:
                new_tmp.next.random = record[m[tmp.random]]
            else:
                new_tmp.next.random = None
            tmp, new_tmp = tmp.next, new_tmp.next
        return new_head.next